### PR TITLE
front: add stop type in stdcm

### LINKED ISSUE #8681: Add selector for stop type in STDCM
The selector must appear after a PR has been selected.

- [x] Each milestone (excluding origin and destination) has a selector for the type of stop. Single choice among the 3 following choices :
    - Service stop
    - Passage (`stop_for = None`)
    - Driver lift
- [x] The “Passage stop” type is selected by default when a PR is added.
- [x] For all intermediate waypoints, the user cannot set the time for the waypoint: “As soon as possible” is selected.
- [x] If the user selects the “Service stop” or “Driver switch” type, the field for entering the duration of the stop is displayed (mandatory).
- [x] If “Driver switch” is selected, the default stop time is set to 3 minutes. This time can be modified by the user, with a warning if it falls below 3mn.
- [x] In case of “Service stop” type, the stop time is not pre-filled.

Image

### PR DESCRIPTION
**Pull request by Akctarus**
closes #8681

### CODE REVIEW

**Comment by Caracol3** [comment]
Nice work ! :-) 
I spotted one thing. When you  have selected 'service stop' and then 'driver switch' or the opposite, there is no reset of the duration. I think we should reajust the duration of the default stop when changing a type.

File: front/src/applications/stdcmV2/components/StdcmVias.tsx
Comment (SharglutDev) [inline front/src/applications/stdcmV2/components/StdcmVias.tsx:1]: Not coming from your PR but there is a warning for a missing `key`. We probably should convert the `fragment` l.88 to a `div` and move the key l.97 inside

File: front/public/locales/en/stdcm.json [hunk 1]
```diff
@@ -77,8 +77,15 @@
     "origin": "Origin",
     "preciseTime": "precise time",
     "stopFor": "Minimum stop time",
+    "stopType": {
+      "driverSwitch": "driver switch",
+      "passageTime":"passage time",
```
Comment (SharglutDev) [inline front/public/locales/en/stdcm.json]:
```suggestion
      "passageTime": "passage time",
```

File: front/src/styles/scss/applications/stdcmV2/_card.scss [hunk 2]
```diff
@@ -49,21 +49,29 @@
   }
 
   .stdcm-v2-card__body {
-    padding: 14px 21px 21px 21px;
+    padding: 14px 16px 21px 16px;
```
Comment (SharglutDev) [inline front/src/styles/scss/applications/stdcmV2/_card.scss]: Padding seems wrong here for the default card (see mockup below) and we should use rem. Maybe there should a specific class for this component. I'm fine if it's done in another PR as it doesn't concern yours but it would be great if you could create the issue and add it to the Meta interface stdcm in confidentials :)

Capture d’écran 2024-09-24 à 22 44 16

File: front/src/styles/scss/applications/stdcmV2/_card.scss
Comment (SharglutDev) [inline front/src/styles/scss/applications/stdcmV2/_card.scss:1]: I know it wasn't strictly follow in stdcm but we try to use `rem` instead of `px`. If you could convert at least the one you changed here :) (if you don't have it there is a nice vs code extension : "px to rem" to help you)

File: front/src/styles/scss/applications/stdcmV2/_card.scss [hunk 3]
```diff
@@ -49,21 +49,29 @@
...
     }
 
     .feed-back {
-      padding-left: 0rem;
       padding-right: 0.3rem;
     }
 
+    .selector {
+      padding-left: 3px;
+      padding-right: 4px;
```
Comment (SharglutDev) [inline front/src/styles/scss/applications/stdcmV2/_card.scss]: You can use `padding-inline` instead of these too

File: front/src/applications/stdcmV2/components/StdcmOpSchedule.tsx [hunk 4]
```diff
@@ -85,50 +85,46 @@ const StdcmOpSchedule = ({
...
+            onArrivalTypeChange(e as ArrivalTimeTypes);
+          }
+        }}
+        options={['preciseTime', 'asSoonAsPossible']}
+        getOptionLabel={(option) => t(`trainPath.${option}`)}
+        getOptionValue={(option) => option}
+        disabled={disabled}
+      />
       {opScheduleTimeType === 'preciseTime' && (
         <div className="d-flex">
```
Comment (SharglutDev) [inline front/src/applications/stdcmV2/components/StdcmOpSchedule.tsx:95]: Spacing between the 3 elements doesn't seem to respect to respect the mockup (16px). If it was like that before your changes, could you create an issue add it to the meta in confidential ?

Capture d’écran 2024-09-24 à 23 10 07

File: front/src/applications/stdcmV2/components/StdcmOpSchedule.tsx [hunk 5]
```diff
@@ -85,50 +85,46 @@ const StdcmOpSchedule = ({
...
-              onArrivalTypeChange(e as ArrivalTimeTypes);
-            }
-          }}
-          options={['preciseTime', 'asSoonAsPossible']}
-          getOptionLabel={(option) => t(`trainPath.${option}`)}
-          getOptionValue={(option) => option}
-          disabled={disabled}
-        />
-      </div>
+    <div className="d-flex flex-column selector">
```
Comment (SharglutDev) [inline front/src/applications/stdcmV2/components/StdcmOpSchedule.tsx:80]: The fact that we need to use this new class to adjust element position is not very good and maintainable.

I'm not sure why we have default padding in select and input elements but it's really difficult to integrate them in a complex UI. @kmer2016 do you know the reason ? Maybe we should open an issue for that.

File: front/src/styles/scss/applications/stdcmV2/_card.scss [hunk 6]
```diff
@@ -49,21 +49,29 @@
...
-    padding: 14px 21px 21px 21px;
+    padding: 14px 16px 21px 16px;
     // TODO: remove this style after ui-core component is implemented
     .suggestions > div:first-child {
-      padding-left: 0;
+      padding-left: 8px;
     }
     .suggestions * label,
     .stdcm-v2-via-stop-for * label {
       margin-bottom: 0;
```
Comment (SharglutDev) [inline front/src/styles/scss/applications/stdcmV2/_card.scss:71]: This should be 2px (in rem)

File: front/src/applications/stdcmV2/components/StdcmInputVia.tsx [hunk 7]
```diff
@@ -20,27 +24,68 @@ const StdcmInputVia = ({
...
+    if (isPassageTime && pathStepStopTime !== '0') {
+      newStopTime = '0';
+    } else if (isDriverSwitch) {
+      const stopTimeValue = Number(pathStepStopTime);
+      if (stopTimeValue === 0) {
+        newStopTime = '3';
+      } else if (stopTimeValue < 3) {
+        warning = true;
+      }
+    } else if (isServiceStop && pathStepStopTime <= '0') {
```
Comment (SharglutDev) [inline front/src/applications/stdcmV2/components/StdcmInputVia.tsx]: Why do we check inferior to 0 ?

File: front/src/applications/stdcmV2/components/StdcmInputVia.tsx [hunk 8]
```diff
@@ -20,27 +24,68 @@ const StdcmInputVia = ({
...
+
+    if (isPassageTime && pathStepStopTime !== '0') {
+      newStopTime = '0';
+    } else if (isDriverSwitch) {
+      const stopTimeValue = Number(pathStepStopTime);
+      if (stopTimeValue === 0) {
+        newStopTime = '3';
+      } else if (stopTimeValue < 3) {
+        warning = true;
+      }
```
Comment (SharglutDev) [inline front/src/applications/stdcmV2/components/StdcmInputVia.tsx]: Maybe we can avoid one nested if with we config warning like that at the beginning

```suggestion
   const warning = isDriverSwitch && stopTimeValue < 3
```

File: front/src/applications/stdcmV2/components/StdcmInputVia.tsx [hunk 9]
```diff
@@ -20,27 +24,68 @@ const StdcmInputVia = ({
     pathStep.stopFor ? `${ISO8601Duration2sec(pathStep.stopFor) / 60}` : ''
   );
 
+  const [showWarning, setShowWarning] = useState(false);
```
Comment (SharglutDev) [inline front/src/applications/stdcmV2/components/StdcmInputVia.tsx]: Maybe we could rather have a const like this and avoid having a state for the warning

```suggestion
  const stopWarning = stopType === StdcmStopTypes.DRIVER_SWITCH && Number(pathStepStopTime) < 3;
```

File: front/src/applications/stdcmV2/components/StdcmVias.tsx [hunk 10]
```diff
@@ -27,10 +29,19 @@ const StdcmVias = ({ disabled = false, setCurrentSimulationInputs }: StdcmConfig
...
+  );
+
   const intermediatePoints = useMemo(() => pathSteps.slice(1, -1), [pathSteps]);
 
   const updatePathStepsList = (pathStep: PathStep | null, index: number) => {
-    const newPathSteps = replaceElementAtIndex(pathSteps, index, pathStep);
+    if (!pathStep) return;
+    const updatedPathStep = {
+      ...pathStep,
+    };
```
Comment (SharglutDev) [inline front/src/applications/stdcmV2/components/StdcmVias.tsx]: Why do we need that ?

File: front/src/applications/stdcmV2/components/StdcmVias.tsx [hunk 11]
```diff
@@ -27,10 +29,19 @@ const StdcmVias = ({ disabled = false, setCurrentSimulationInputs }: StdcmConfig
     useOsrdConfActions() as StdcmConfSliceActions;
   const pathSteps = useSelector(getPathSteps);
 
+  const [stopTypes, setStopTypes] = useState<StdcmStopTypes[]>(
+    pathSteps.map(() => StdcmStopTypes.PASSAGE_TIME)
+  );
```
Comment (SharglutDev) [inline front/src/applications/stdcmV2/components/StdcmVias.tsx]: It might be safer to group them by pathStep ids instead of their indexes in case one of the via is removed ? What do you think ?

```suggestion
  const [stopTypes, setStopTypes] = useState<Record<string, StdcmStopTypes>>(
    compact(pathSteps).reduce(
      (acc, cur) => {
        acc[cur.id] = StdcmStopTypes.PASSAGE_TIME;
        return acc;
      },
      {} as Record<string, StdcmStopTypes>
    )
  );
```

File: front/src/applications/stdcmV2/components/StdcmInputVia.tsx [hunk 12]
```diff
@@ -20,27 +24,68 @@ const StdcmInputVia = ({
...
+      newStopTime = '1';
+    }
+
+    if (newStopTime !== pathStepStopTime) {
+      setPathStepStopTime(newStopTime);
+      updatePathStepStopTime(newStopTime);
+    }
+
+    setShowWarning(warning);
+  }, [stopType, pathStepStopTime]);
```
Comment (SharglutDev) [inline front/src/applications/stdcmV2/components/StdcmInputVia.tsx]: Not sure it's healthy to have in the dependency array the state and update with it's setter inside, it might go through some loops

File: front/src/applications/stdcm/components/SimulationReportSheet.tsx [hunk 13]
```diff
@@ -134,14 +134,18 @@ const SimulationReportSheet = ({
                   <View style={styles.convoyAndRoute.stopTableStartWidth}>
                     <TD>{t('startStop')}</TD>
                   </View>
-                  <View style={styles.convoyAndRoute.stopTableMotifWidth}>
-                    <TD>{t('motif')}</TD>
+                  <View style={styles.convoyAndRoute.stopTableStopTypeWidth}>
+                    <TD>{t('stopType')}</TD>
```
Comment (SharglutDev) [inline front/src/applications/stdcm/components/SimulationReportSheet.tsx:139]: The stop type should be capitalized

**Review by SharglutDev — COMMENTED** [review]
Thank you for the PR, left some comments

File: front/public/locales/en/stdcm.json (hunk 1 above)
Comment (Akctarus) [reply front/public/locales/en/stdcm.json]: done

File: front/src/styles/scss/applications/stdcmV2/_card.scss (hunk 2 above)
Comment (Akctarus) [reply front/src/styles/scss/applications/stdcmV2/_card.scss]: Fixed, it's a little hard since we can't see the arrow size

File: front/src/styles/scss/applications/stdcmV2/_card.scss (hunk 3 above)
Comment (Akctarus) [reply front/src/styles/scss/applications/stdcmV2/_card.scss]: ty ! done

File: front/src/applications/stdcmV2/components/StdcmOpSchedule.tsx (hunk 5 above)
Comment (Akctarus) [reply front/src/applications/stdcmV2/components/StdcmOpSchedule.tsx:80]: I agree, but do I leave it like that until I find a better solution?

File: front/src/styles/scss/applications/stdcmV2/_card.scss (hunk 6 above)
Comment (Akctarus) [reply front/src/styles/scss/applications/stdcmV2/_card.scss:71]: We need to have a 24px padding between edge of card and input, we already have a 16px padding with another class, so 8px (0.5rem) is good

File: front/src/applications/stdcmV2/components/StdcmInputVia.tsx (hunk 9 above)
Comment (Akctarus) [reply front/src/applications/stdcmV2/components/StdcmInputVia.tsx]: fixed

File: front/src/applications/stdcmV2/components/StdcmInputVia.tsx (hunk 8 above)
Comment (Akctarus) [reply front/src/applications/stdcmV2/components/StdcmInputVia.tsx]: fixed

File: front/src/applications/stdcmV2/components/StdcmVias.tsx (hunk 10 above)
Comment (Akctarus) [reply front/src/applications/stdcmV2/components/StdcmVias.tsx]: an oversight, well seen, fixed

File: front/src/applications/stdcm/components/SimulationReportSheet.tsx (hunk 13 above)
Comment (Akctarus) [reply front/src/applications/stdcm/components/SimulationReportSheet.tsx:139]: fixed

File: front/src/styles/scss/applications/stdcmV2/_card.scss
Comment (Akctarus) [reply front/src/styles/scss/applications/stdcmV2/_card.scss:1]: done !

File: front/src/applications/stdcmV2/components/StdcmOpSchedule.tsx (hunk 4 above)
Comment (Akctarus) [reply front/src/applications/stdcmV2/components/StdcmOpSchedule.tsx:95]: Moved to META  : https://github.com/OpenRailAssociation/osrd/issues/8991

File: front/src/applications/stdcmV2/components/StdcmVias.tsx (hunk 11 above)
Comment (Akctarus) [reply front/src/applications/stdcmV2/components/StdcmVias.tsx]: done

File: front/src/applications/stdcmV2/components/StdcmInputVia.tsx (hunk 7 above)
Comment (Akctarus) [reply front/src/applications/stdcmV2/components/StdcmInputVia.tsx]: Changed

File: front/src/applications/stdcmV2/components/StdcmInputVia.tsx (hunk 12 above)
Comment (Akctarus) [reply front/src/applications/stdcmV2/components/StdcmInputVia.tsx]: I had to keep it but I moved lot of the logic in parent file.

**Comment by Caracol3** [comment]
Good job !
I have just one  thing with the behavior of the selector when we hover it after selection.
There is a big arrow instead of our icon.
image

**Comment by Akctarus** [comment]
> Good job ! I have just one thing with the behavior of the selector when we hover it after selection. There is a big arrow instead of our icon. image

It has been fixed in another PR

**Review by Caracol3 — APPROVED** [review]
lgtm (tested)

**Review by SharglutDev — APPROVED** [review]
Lgtm and tested, good job ! Some UI issues remain (not from the PR), they have been added in the meta issue
